### ISSUE TITLE
Intermittent gfx906 GPUVM page faults with Qwen3.8-Flash-Next on 4x MI50

### ISSUE THREAD

**Issue by DByte308** [mxxm-t/mx-llama.cpp#15]
## Summary

On a 4x MI50/gfx906 host, Qwen3.8-Flash-Next intermittently triggers compute-side AMDGPU GPUVM page faults during model load or inference. I reproduced faults under **both layer and tensor split**. Every captured kernel fault was reported on physical GPU `0c:00.0` (normally `ROCm3`, the last rank), but the failure is rare enough that a reversed-rank run did not fault, so physical-card vs logical-rank causality remains unresolved.

This report is based on local commit `62359782533e2cb044c3e85f0dbf628e68449ff2` (`gfx906: pinned host staging buffer for set_tensor`). I noticed current `origin/master` is now `75db261922f9006ca53590b0d7223fb6e4fddf8d`; I have not yet reproduced on that newer head.

## Environment

- Host: Intel X99 / i7-6850K
- GPUs: 4x AMD Vega 20 / MI50, 32 GiB each (`gfx906`)
- PCI addresses: `03:00.0`, `06:00.0`, `09:00.0`, `0c:00.0`
- Kernel: `7.2.0-nobara.p2plow`
- ROCm userspace: 7.14.0
- Fork/image revision: `62359782533e2cb044c3e85f0dbf628e68449ff2`
- Model: `Qwen3.8-Flash-Next-UD-Q4_K_XL`, four GGUF shards, about 105 GiB
- Model architecture: `qwen4exp`

The host uses low-MMIO BAR placement with gfx906's native 44-bit DMA mask. All four 32 GiB BARs are below the mask, KFD reports three peer links per GPU, all 12 directional `hipMemcpyPeer` paths previously passed byte verification, and 4-rank custom AR previously passed output-correctness tests. HBM/PCIe/MMHUB/SDMA/GFX RAS counters remained zero after the faults.

## Representative command

```bash
docker run --name qwen4-layer --shm-size 4g \
  --device /dev/kfd \
  --device /dev/dri/renderD128 \
  --device /dev/dri/renderD129 \
  --device /dev/dri/renderD130 \
  --device /dev/dri/renderD131 \
  --group-add 39 --group-add 105 \
  -e HSA_ENABLE_P2P=1 \
  mx-llama-server:gfx906-qwen4exp \
  --server --host 0.0.0.0 --port 8080 \
  -m /models/Qwen3.8-Flash-Next-UD-Q4_K_XL-00001-of-00004.gguf \
  --device ROCm0,ROCm1,ROCm2,ROCm3 \
  --split-mode layer --no-mmap --gpu-layers 999 \
  --parallel 1 -c 65536 -b 2048 -ub 2048 \
  --spec-type none --cache-ram 8192 --ctx-checkpoints 0
```

Tensor mode additionally used:

```text
HSA_FORCE_FINE_GRAIN_PCIE=1
GGML_CUDA_P2P=1
GGML_ENABLE_CUSTOM_AR=1
GGML_TP_AR_KTIMING=1
```

Layer mode did not enable custom AR or fine-grain PCIe, but did retain `HSA_ENABLE_P2P=1`.

## Observed failures

The failures are intermittent: identical commands can load and complete 2K/8K/32K requests successfully, then fault on another attempt.

Observed locations included:

- model load in `__amd_rocclr_fillBufferAligned`;
- one load while repacking a large tensor on device 2/3;
- inference in `k_bin_bcast` during an 8K prefill with tensor/custom-AR enabled.

All three kernel episodes captured in the same boot pointed to `0c:00.0`:

```text
amdgpu 0000:0c:00.0: [gfxhub0] no-retry page fault (src_id:0 ring:174 vmid:8 pasid:144)
Process llama-server pid 10135 thread llama-server pid 10135
in page starting at address 0x00007f5885661000 from IH client 0x1b (UTCL2)
Faulty UTCL2 client ID: TCP (0x8)
WALKER_ERROR: 0x6
PERMISSION_FAULTS: 0x5
MAPPING_ERROR: 0x1
RW: 0x1
```

Two later examples were:

```text
amdgpu 0000:0c:00.0: [gfxhub0] no-retry page fault ...
Faulty UTCL2 client ID: CPF (0x4)
WALKER_ERROR: 0x6
PERMISSION_FAULTS: 0xb
MAPPING_ERROR: 0x1
RW: 0x0
```

```text
amdgpu 0000:0c:00.0: [gfxhub0] no-retry page fault ...
Faulty UTCL2 client ID: CPC (0x5)
WALKER_ERROR: 0x6
PERMISSION_FAULTS: 0x3
MAPPING_ERROR: 0x1
RW: 0x0
```

The `TCP`/`CPF`/`CPC` clients and the observed kernel locations make these look compute/command-side rather than a simple failed SDMA peer copy.

## Controls performed

- Layer and tensor split have both faulted.
- Reversing the device list to `ROCm3,ROCm2,ROCm1,ROCm0` completed one load plus three requests without a fault. Because the fault is rare, this was inconclusive rather than evidence that reversal fixes it.
- After cleanup, roughly nine additional model loads completed without a fault.
- Multiple subsequent runs with `--cache-ram 0` were clean, but the event rate is too low to claim this is a fix.
- All RAS error counters remained zero.
- Current BARs/topology are healthy, and prior exhaustive P2P copy/custom-AR correctness tests passed.

## Possibly relevant local delta

Commit `623597825` adds one process-global pinned staging allocation in `ggml-cuda.cu`:

```cpp
static uint8_t * ggml_cuda_pinned_staging(size_t need) {
    static uint8_t * buf = nullptr;
    static size_t cap = 0;
    if (need > cap) {
        if (buf) cudaFreeHost(buf);
        cudaMallocHost((void **) &buf, need);
        cap = need;
    }
    return buf;
}
```

The buffer is shared by all devices/callers and is not guarded by a mutex or made thread-local/per-device. I have **not proven** that this is the cause, but concurrent set/get paths could overwrite it or resize/free it while another operation uses it. It is also possible that the fault is elsewhere in graph-buffer lifetime/rebinding or qwen4exp's multi-GPU path.

Current master also contains later allocation/staging changes (`54702a718`, `20af9a480`, `27f755681`) that are absent from this image, so retesting a current-master build is the next step.

## Separate non-fault finding (ruled out as truncation)

An apparent deterministic "one-token truncation" at one 8K test prompt was investigated separately. The native `/completion` response showed:

```text
tokens_predicted=1
stop_type=eos
sampled token id=248046
```

The first token was sampled directly from the prefill logits and was an empty EOG token. Its log probability was only modestly above `<thinking>` and content tokens. It reproduced with reversed device order and changed with prompt content/length. This is an immediate model EOS, **not** the server skipping decode, and appears independent of the intermittent GPUVM fault.

## Questions

1. Are there known qwen4exp/meta-backend buffer-lifetime issues that could produce these compute-side GPUVM faults on the last rank?
2. Should the pinned staging buffer introduced in `623597825` be per-device/thread-local or guarded across the full copy lifetime?
3. Which current-master tracing options would best identify ownership of the faulting `0x7f...` virtual address and the tensor/kernel using it?

I can rebuild/retest current master and add tracing if you can suggest the most useful toggles or instrumentation.

**Comment by DByte308** [comment]
## Severity update: sustained run caused a full host hang/reboot

After the initial report, a sustained multi-seed 4-GPU decode run hard-froze the host. The machine subsequently rebooted (previous boot ended at 20:40:25; current boot began at 20:41:40).

The previous boot's kernel journal confirms this was an AMDGPU recovery deadlock rather than only a userspace server failure. Representative evidence:

```text
INFO: task kworker/1:1:61031 blocked for more than 245 seconds.
Workqueue: events_freezable svm_range_restore_work [amdgpu]
...
svm_range_list_lock_and_flush_work [amdgpu]
svm_range_restore_work [amdgpu]
...
INFO: task kworker/1:1:61031 <writer> blocked on an rw-semaphore likely owned by task llama-server:85195 <reader>

INFO: task kworker/u48:14:72077 blocked for more than 245 seconds.
Workqueue: amdgpu-reset-dev drm_sched_job_timedout [gpu_sched]
...
dm_suspend [amdgpu]
amdgpu_device_pre_asic_reset [amdgpu]
amdgpu_device_asic_reset [amdgpu]
amdgpu_device_gpu_recover [amdgpu]
amdgpu_job_timedout [amdgpu]

INFO: task kworker/u49:16:73626 blocked for more than 122 seconds.
Workqueue: ttm ttm_bo_delayed_delete [ttm]
...
dma_fence_wait_timeout
ttm_bo_delayed_delete [ttm]
```

Before this sustained run, approximately nine lighter/cache-ram-0 loads had completed cleanly, emphasizing that the failure is intermittent and load-sensitive rather than resolved by those clean trials.

I am pausing further sustained 4-GPU reproduction on this build because the failure can wedge GPU reset/SVM/TTM strongly enough to take down the host. Production recovered after reboot and is healthy.

The completed A/B report is retained locally at:

```text
/mnt/nvme1/llaminar-test/qwen4-ab/AB-QWEN4-LAYER-vs-TENSOR-2026-09-18.md
```

Its performance headline (before the freeze) was layer prefill about 1.6-2.1x tensor prefill at medium/long context; generation was roughly 21-35 tok/s but confounded by prompt-dependent immediate EOG samples.

**Comment by DByte308** [comment]
## Current-master retest: fault persists; global pinned-staging hypothesis ruled out

Rebuilt and tested current fork master `75db261922f9006ca53590b0d7223fb6e4fddf8d` as `gfx906-qwen4exp-75db` using a stop-at-first-failure gate:

1. Bare model load: PASS
2. Fresh load + 2K request: PASS (request immediately sampled model EOG, unrelated)
3. Fresh load for the 8K stage: **FAILED during load** with the same intermittent fault:

```text
Memory access fault by GPU node-4 ...
Reason: Page not present
```

The failure occurred while q8 repack was active (log had reached repacking on device 2), and the kernel again attributed the fault to the last physical GPU/rank (`0c:00.0` mapping in the normal device order).

This build does **not** contain local commit `623597825` or its process-global `ggml_cuda_pinned_staging` buffer. Therefore that global pinned-staging implementation is ruled out as the cause of this load-time fault.

The next active discriminator is a current-master `--no-repack` gate:

- stable without repack would localize the fault to repack upload/scratch/asynchronous lifetime;
- a fault without repack would point instead toward the meta allocator/buffer lifetime or physical `0c:00.0` card/slot path.

No sustained testing is being performed; each stage stops after the first anomaly because the previous build caused an unrecoverable AMDGPU reset/SVM deadlock.

**Comment by DByte308** [comment]
## `--no-repack` retest: same load-time fault

The current-master (`75db26192`) stop-at-first-failure gate was repeated with `--no-repack`:

1. Stage A bare load: PASS
2. Stage B fresh load: **FAILED during model load** with the same fault:

```text
Memory access fault by GPU node-4 ...
Reason: Page not present
address 0x7f...
```

Therefore both the q8 repack path and the old process-global pinned-staging patch are excluded as necessary causes. The failure reproduces on current master with repack disabled and continues to target the physical last-rank mapping (`0c:00.0` in normal order).

Further 4-GPU Qwen load/stress testing is paused because an earlier sustained run deadlocked AMDGPU SVM/reset/TTM and froze the host.

The remaining leading branches are:

1. meta-backend allocator/rebind/tensor-lifetime handling associated with the last rank; or
2. the physical `0c:00.0` card, slot/root-port, power path, or display ownership.

The next discriminating test should be a physical card/slot swap rather than more repeated loads. If the fault follows the card, investigate the GPU/VBIOS/power; if it stays at `0c:00.0`, investigate that slot/root port/display ownership; if it follows logical last rank after remapping, instrument meta allocator ownership and buffer lifetimes.

Gate scripts and logs are retained under:

```text
/mnt/nvme1/llaminar-test/qwen4-ab/
```

**Comment by evilJazz** [comment]
I've seen this issue before, but only when cache is on and contexts are restored or speculative decoding (even ngram-mod) is turned on. https://github.com/ggml-org/llama.cpp/issues/28425#issuecomment-5575209923
Now back on mixa3607/rocm-gfx906:7.2.4-complete and have this patch applied on generic llama-cpp: https://github.com/ggml-org/llama.cpp/pull/27405 which seems to help with the cache/context issue. The patch also fixes the issue on 5xR9700, so not MI50 exclusive and not a hardware/memory issue.

**Comment by DByte308** [comment]
## Resolved: display/compositor co-use of the display GPU (0c:00.0) was the trigger

**Update 2026-09-19.** The root-cause discriminator has been identified and the fault is no longer reproducible. Summary of what we found and what closing this on:

### The decisive capture
In a clean A/B on commit `75db26192` (plain, uninstrumented image), the kernel journal caught the faulting **process** on the repro:

```text
amdgpu 0000:0c:00.0: [gfxhub0] no-retry page fault (src_id:0 ring:190 vmid:5 pasid:86)
Process kwin_wayland pid 1741 thread kwin_wayla:cs0 pid 1853
in page starting at address 0x000080010c677000 from IH client 0x1b (UTCL2)
Faulty UTCL2 client ID: CB (0x0)  ... then  TCP (0x8)
```

Not `llama-server` this time — **`kwin_wayland`**, the Wayland desktop compositor, faulting on the **display GPU**. `0c:00.0` is `renderD131`/`card5`, the **only** card with a live connected DP output — i.e. the desktop display. Llama's 4-GPU load uses that same card as `ROCm3` while the compositor renders on it, so every fault landed on `0c:00.0` regardless of split/device-order.

### Controlled A/B (same hardware, same session)
- Compositor **on** `0c:00.0`   : load1 clean, **load2 fault** (kwin, gfxhub0 CB/TCP)
- Compositor **off** `0c:00.0`  (output moved to VKMS virtual display + monitor removed; `kwin_wayland` no longer holds `renderD131`): **3/3 clean** cold loads
- Then a **full clean tensor-vs-layer A/B** (both splits, 2K/8K/32K): **0 faults**, kernel journal clean throughout; results appended to `AB-QWEN4-LAYER-vs-TENSOR-2026-09-18.md`.

### What this rules in/out
- **Not** a bug in the fork's repack path or the old `623597825` pinned-staging buffer (both already ruled out by earlier `75db` and `--no-repack` gates).
- The ROCm env-staging H2D workaround (`GPU_PINNED_MIN_XFER_SIZE`/`GPU_STAGING_BUFFER_SIZE`) did **not** prevent the kwin-attributed fault — it was not a llama async-H2D fault, so it neither proves nor disproves ROCm rocm-systems#4817 for llama-owned copies.
- This is an **environment/driver/display contention trigger**, not a defect in mx-llama.cpp. It also corroborates @evilJazz's point that this is neither MI50-exclusive nor a memory/hardware issue.

### Caveat
The earlier captures in this thread were attributed to `Process llama-server` with compute-side clients (TCP/CPF/CPC), so I am not claiming every historical fault was literally the compositor's. But the environmental change that removes the compositor from the display card stopped the reproduction entirely and let 4-GPU loads run clean, so it is the strongest and actionable trigger we found.

### Resolution / mitigation (why it's fixed)
**Keep the desktop/compositor from rendering on `0c:00.0` during multi-GPU loads** — run them headless, on the VKMS virtual output, or with the display on a different card. With that in place, 4-GPU Qwen3.8-Flash-Next loads are stable and fault-free.

Reopen only if the fault reproduces **while the compositor is already off `renderD131`** — that would point back at a card/slot/hardware or code-lifetime path.

**Comment by DByte308** [comment]
Closing: reproduced root cause was desktop/Wayland compositor (kwin_wayland) faulting on the display GPU 0c:00.0 under concurrent 4-GPU load. Mitigation (keep the compositor off that card / run headless) yields clean bounded loads and a full fault-free A/B. Not a code defect in mx-llama.cpp. Reopen if it recurs with the compositor already off /dev/dri/renderD131.
